### DISCUSSION TITLE
404 when running on apache

### DISCUSSION THREAD

**Discussion by sebiko3** [clue/framework-x#111] [Q&A] [answered]
Hi there,

I just tried to use framework-x but got stopped right away. 
using the sample quickstart code:

```php
  require __DIR__ . '/vendor/autoload.php';

  $app = new FrameworkX\App();

  $app->get('/', function () {
    return new React\Http\Message\Response(
      200,
      [],
      "Hello wörld!\n"
    );
  });

  $app->run();
```

I only get 404 when I try to access it on my local (subdirectory) - however using the builtin webserver it works.
Any ideas why?

Thanks in advance.

P.S.: I havn't used a public dir instead just put the index.php into the docroot, shouldn't be an issue or is there some internal requirement to serve from public?

**Comment by SimonFrings** [comment, marked as answer]
Hey @p13rnd, thanks for bringing this up :+1: 

Running your code on apache needs some additional setup which is not necessary when working with the built-in webserver. One difference is to adjust your `.htaccess` file when using apache. You can read all about this in [our documentation](https://framework-x.org/docs/best-practices/deployment/#apache). The `.htaccess` file looks something like this:

```
RewriteEngine On

RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule .* index.php
```
There is no requirement to add a public dir, but we always recommend doing it. In our cases the public folder functions as our document root. Only files, where it's not critical if other people can access them, should be in there, for example if you have an `.env` file, it's rather dangerous to put it inside the public folder for the world to see ^^

I hope this helps, if you have any more questions let us know :+1:

**Reply by sebiko3** [reply]
Hi @SimonFrings 

just another thing to mention in this regard: using the built-in webserver is fine to me as well, however, I have to restart it every time I make changes to the source. Fine for prod environments but for dev it's somewhat of a hassle. Is there any option to auto-reload/watch-for-changes? 

Thanks. <3

**Reply by SimonFrings** [reply]
Hey @magicbyt3 

Not for the built-in web server at the moment. This has also been addressed in #86, you should find some answers in there. As @clue says:

> At the moment, the easiest work around is to use PHP's development web server for development purposes and the built-in web server for production use:
> ```
> # development (reloads source for every request)
> $ php -S localhost:8080 public/index.php
> 
> # deployment (keeps everything in memory)
> $ php public/index.php
> ```

> Longer term, we're definitely interested in bringing in hot reloading also for the built-in web server.

I hope this helps.

**Comment by sebiko3** [comment]
Ah got it! I've missed the production deployment part of the docs as my intention was to only try the framework first. You may want to add this part to the quickstart section so people won't miss it and get turned off. 

For the public dir part, I am well aware of the security constraints but havn't been entirely sure if there are any other framework enforced conventions. 

Thanks!

**Comment by SimonFrings** [comment]
@p13rnd You're right, there should be a reference somewhere inside the quickstart section, thanks for pointing that out :+1:

I've already submitted a pull request: #112
